### PR TITLE
gpt4all-chat: init at 2.6.2

### LINKED ISSUE #259798: Package request: GPT4All
**Project description**

Open-source assistant-style large language models that run locally on your CPU

GPT4All is an open-source ecosystem that allows you to integrate Large Language Models (LLMs) into applications without the need for a platform or hardware subscription. 

**Metadata**

* homepage URL: https://gpt4all.io/index.html
* source URL: https://github.com/nomic-ai/gpt4all
* license: mit
* platforms: linux, windows, darwin

### PR DESCRIPTION
**Pull request by drupol**
Source: https://github.com/nomic-ai/gpt4all

Disclaimer: NIH, shamelessly stolen from @polygon's flake at https://github.com/polygon/gpt4all-nix, all the props to him/her! @polygon if you want to claim commit authorship, feel free to amend this PR and do the necessary changes, I'm totally OK with that.

Fix #259798

Tested locally successfully:

image

image

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by drupol** [comment]
@GrahamcOfBorg build gpt4all-chat

**Comment by drupol** [comment]
Marking it as draft, I just noticed that there's a `.desktop` file in the source, going to add it.

File: pkgs/by-name/gp/gpt4all-chat/package.nix [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    "-DKOMPUTE_OPT_DISABLE_VULKAN_VERSION_CHECK=ON"
+    "-DKOMPUTE_OPT_USE_BUILT_IN_FMT=OFF"
+  ];
+
+  postInstall = ''
+    mv $out/bin/chat $out/bin/gpt4all-chat
+  '';
+
+  meta = {
+    description = "Gpt4all";
```
Comment (polygon) [inline pkgs/by-name/gp/gpt4all-chat/package.nix:62]: Maybe add the description from their webpage here:

A free-to-use, locally running, privacy-aware chatbot. No GPU or internet required.

**Review by polygon — COMMENTED** [review]
Looks good to me, only thing I'd change is the description.

**Comment by polygon** [comment]
I also noticed that there is an `include` folder in the outputs. I don't believe that we actually need that one.

File: pkgs/by-name/gp/gpt4all-chat/package.nix (hunk 1 above)
Comment (drupol) [reply pkgs/by-name/gp/gpt4all-chat/package.nix]: Done!

**Comment by drupol** [comment]
I don't really know why CI is not building the new derivation, so I will trigger it from here.

@GrahamcOfBorg build gpt4all-chat

**Comment by drupol** [comment]
I don't really know why CI is not building the new derivation, so I will trigger it from here.

@GrahamcOfBorg build gpt4all-chat

**Comment by drupol** [comment]
I don't really know why CI is not building the new derivation, so I will trigger it from here.

@GrahamcOfBorg build gpt4all-chat

**Review by polygon — COMMENTED** [review]
I noticed two other things looking at the output:

* There's a few static archives in the `lib` folder ending in `.a` which I think can be removed. This should reduce the closure size by more than 50%.
* There seem to be shared libraries ending in `default` and `avxonly`. Maybe they now build both by default and load the right one at runtime. Hence, we can consider removing the `withAvx2` option. But I have no machine old enough right now to test this for sure.

**Comment by drupol** [comment]
Sadly, removing the `lib` folder broke the app.

**Comment by polygon** [comment]
> Sadly, removing the `lib` folder broke the app.

Not the whole folder, just the files ending in `.a` inside the folder. Just gave it a got and seems to work without issues.

**Comment by drupol** [comment]
That's indeed working when removing `lib/*.a` files! Committed and pushed.

File: pkgs/by-name/gp/gpt4all-chat/package.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    qt6.qthttpserver
+    qt6.qtwebengine
+    qt6.qt5compat
+    shaderc
+    vulkan-headers
+    wayland
+  ];
+
+  cmakeFlags = [
+    "-DGPT4ALL_AVX_ONLY=ON"
```
Comment (Tungsten842) [inline pkgs/by-name/gp/gpt4all-chat/package.nix:50]: This environment variable isn't utilized:
```
Manually-specified variables were not used by the project:
    GPT4ALL_AVX_ONLY
```

**Comment by drupol** [comment]
I don't really know why CI is not building the new derivation, so I will trigger it from here.

@GrahamcOfBorg build gpt4all-chat

**Review by Tungsten842 — CHANGES_REQUESTED** [review]

**Comment by Tungsten842** [comment]
In the device selection in cannot choose the gpu in any way, even though vulkan is included in the package. The only option avaliable is the cpu.

File: pkgs/by-name/gp/gpt4all-chat/package.nix (hunk 2 above)
Comment (drupol) [reply pkgs/by-name/gp/gpt4all-chat/package.nix]: Fixed! Thanks!

**Comment by drupol** [comment]
> In the device selection in cannot choose the gpu in any way, even though vulkan is included in the package. The only option avaliable is the cpu.

Do you have any idea on how to fix that?

**Comment by Tungsten842** [comment]
> > In the device selection in cannot choose the gpu in any way, even though vulkan is included in the package. The only option avaliable is the cpu.
> 
> Are you running NixOS or Nix on another distribution?

NixOS

**Comment by Tungsten842** [comment]
> > In the device selection in cannot choose the gpu in any way, even though vulkan is included in the package. The only option avaliable is the cpu.
> 
> Do you have any idea on how to fix that?

Not really

**Comment by drupol** [comment]
I don't really know why CI is not building the new derivation, so I will trigger it from here.

@GrahamcOfBorg build gpt4all-chat

**Comment by dingesista** [comment]
Hello ! I get this error

ix run github:NixOS/nixpkgs/nixos-unstable#gpt4all-chat
[Warning] (Tue Feb 20 10:41:30 2024): Detected locale "C" with character encoding "ANSI_X3.4-1968", which is not UTF-8.
Qt depends on a UTF-8 locale, and has switched to "C.UTF-8" instead.
If this causes problems, reconfigure your locale. See the locale(1) manual
for more information.
[Debug] (Tue Feb 20 10:41:31 2024): deserializing chats took: 0 ms
[Warning] (Tue Feb 20 10:41:31 2024): Could not connect "org.freedesktop.NetworkManager" to "stateChanged" : Type not registered with QtDBus in parameter list: QNetworkManagerInterface::NMState
[Warning] (Tue Feb 20 10:41:31 2024): Could not connect "org.freedesktop.NetworkManager" to "connectivityChanged" : Type not registered with QtDBus in parameter list: QNetworkManagerInterface::NMConnectivityState
[Warning] (Tue Feb 20 10:41:31 2024): Could not connect "org.freedesktop.NetworkManager" to "deviceTypeChanged" : Type not registered with QtDBus in parameter list: QNetworkManagerInterface::NMDeviceType
[Warning] (Tue Feb 20 10:41:31 2024): Could not connect "org.freedesktop.NetworkManager" to "meteredChanged" : Type not registered with QtDBus in parameter list: QNetworkManagerInterface::NMMetered
[Warning] (Tue Feb 20 10:41:31 2024): qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QColorSpace(), profile  QSurfaceFormat::NoProfile)
[Warning] (Tue Feb 20 10:41:31 2024): qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QColorSpace(), profile  QSurfaceFormat::NoProfile)
[Warning] (Tue Feb 20 10:41:31 2024): qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QColorSpace(), profile  QSurfaceFormat::NoProfile)
[Fatal] (Tue Feb 20 10:41:31 2024): Could not initialize GLX
Aborted


and this one 
nix run github:NixOS/nixpkgs/nixpkgs-unstable#gpt4all-chat
[Warning] (Tue Feb 20 10:42:32 2024): Detected locale "C" with character encoding "ANSI_X3.4-1968", which is not UTF-8.
Qt depends on a UTF-8 locale, and has switched to "C.UTF-8" instead.
If this causes problems, reconfigure your locale. See the locale(1) manual
for more information.
[Debug] (Tue Feb 20 10:42:32 2024): deserializing chats took: 1 ms
[Warning] (Tue Feb 20 10:42:32 2024): Could not connect "org.freedesktop.NetworkManager" to "stateChanged" : Type not registered with QtDBus in parameter list: QNetworkManagerInterface::NMState
[Warning] (Tue Feb 20 10:42:32 2024): Could not connect "org.freedesktop.NetworkManager" to "connectivityChanged" : Type not registered with QtDBus in parameter list: QNetworkManagerInterface::NMConnectivityState
[Warning] (Tue Feb 20 10:42:32 2024): Could not connect "org.freedesktop.NetworkManager" to "deviceTypeChanged" : Type not registered with QtDBus in parameter list: QNetworkManagerInterface::NMDeviceType
[Warning] (Tue Feb 20 10:42:32 2024): Could not connect "org.freedesktop.NetworkManager" to "meteredChanged" : Type not registered with QtDBus in parameter list: QNetworkManagerInterface::NMMetered
[Warning] (Tue Feb 20 10:42:32 2024): qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QColorSpace(), profile  QSurfaceFormat::NoProfile)
[Warning] (Tue Feb 20 10:42:32 2024): qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QColorSpace(), profile  QSurfaceFormat::NoProfile)
[Warning] (Tue Feb 20 10:42:32 2024): qglx_findConfig: Failed to finding matching FBConfig for QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize -1, redBufferSize 1, greenBufferSize 1, blueBufferSize 1, alphaBufferSize -1, stencilBufferSize -1, samples -1, swapBehavior QSurfaceFormat::SingleBuffer, swapInterval 1, colorSpace QColorSpace(), profile  QSurfaceFormat::NoProfile)
[Fatal] (Tue Feb 20 10:42:32 2024): Could not initialize GLX
Aborted

**Comment by drupol** [comment]
The solution was to do:

```
NIXPKGS_ALLOW_UNFREE=1 nix run --override-input nixpkgs nixpkgs/nixos-unstable --impure github:guibou/nixGL -- nix run nixpkgs#gpt4all-chat
```

**Comment by dingesista** [comment]
Correct, ack, thanks!

**Pull request merged**